Sources/Bow/Typeclasses/Monad.swift

```diff
@@ -1,50 +1,135 @@
 import Foundation
 
+/// A Monad provides functionality to sequence operations that are dependent from one another.
+///
+/// Instances of `Monad` must obey the following rules:
+///
+///     flatMap(pure(a), f) == f(a)
+///     flatMap(fa, pure) == fa
+///     flatMap(fa) { a in flatMap(f(a), g) } == flatMap(flatMap(fa, f), g)
+///
+/// Also, instances of `Monad` derive a default implementation for `Applicative.ap` as:
+///
+///     ap(ff, fa) == flapMap(ff, { f in map(fa, f) }
 public protocol Monad: Applicative {
+    /// Sequentially compose two computations, passing any value produced by the first as an argument to the second.
+    ///
+    /// - Parameters:
+    ///   - fa: First computation.
+    ///   - f: A function describing the second computation, which depends on the value of the first.
+    /// - Returns: Result of composing the two computations.
     static func flatMap<A, B>(_ fa: Kind<Self, A>, _ f: @escaping (A) -> Kind<Self, B>) -> Kind<Self, B>
+
+    /// Monadic tail recursion.
```

Comment:
In this comment I'd mention the reasoning behind this method and the paper that initially introduced it.

```suggestion
    /// Monadic tail recursion.
    ///
    /// tailRecM can be used for computations that can potentially make the stack overflow.
    ///
    /// initially introduced in Stack Safety for Free: https://functorial.com/stack-safety-for-free/index.pdf
```

Also, I'd mention that there's a default implementation for this, but it's not always lawful: (in scala, IDK any swift)

```scala
  def tailRecM[A, B](a: A)(f: A => F[Either[A, B]]): F[B] =
    flatMap(f(a)) {
      case Right(b) => pure(b)
      case Left(nextA) => tailRecM(nextA)(f)
    }
```

Reply:
We haven't made a default implementation, but it is something that I'd consider to add in the future.